dynamic_programming/house_robber.cpp

```diff
@@ -0,0 +1,104 @@
+/**
+ * @file
+ * @brief Implementation of [House Robber Problem] algorithm
+ * @details
+ * Solution of House robber problem uses a dynamic programming concept that works in \f$O(n)\f$ time and works in \f$O(1)\f$ space.
+ */
+
+#include <cassert>    /// for assert
+#include <iostream>   /// for io operations
+#include <vector>     /// for std::vector
+
+/**
+ * @namespace dynamic_programming
+ * @brief House robber algorithm
+ */
+namespace dynamic_programming {
+/**
+ * @namespace house_robber
+ * @brief Functions for House Robber algorithm
+ */
+namespace house_robber {
+/**
+ * @brief The main function implements house robber algorithm using dynamic programming
+ * @param money array containing money in the ith house
+ * @param n size of array
+ * @returns maximum amount of money that can be robbed
+ */
+std::uint64_t houseRobber(const std::vector<int> &money, int n) {
+
+    if (n == 0) { // if there is no house
+      return 0;
+    }
+    if (n == 1) { // if there is only one house
+      return money[0];
+    }
+    if (n == 2) { // if there are two houses, one with the maximum amount of money will be robbed
+      if (money[0] > money[1]) {
+        return money[0];
+      }
+      return money[1];
+    }
+    int max_value; // contains maximum stolen value at the end
+    int value1 = money[0];
+    int value2;
+    if (money[0] > money[1]) {
+        value2 = money[0];
+    }
+    else{
+        value2 = money[1];
+    }
+    for (int i=2; i<n; i++) {
+       if (money[i]+value1 > value2) {
+          max_value = (money[i]+value1);
+       }
+       else{
+          max_value = (value2);
+       }
+       value1 = value2;
+       value2 = max_value;
```

Comment:
Similarly you can think how to use std::max() here as well. Code will become more cleaner.